Tank Desant: Rideable Tank And Overhauled Interactions

SunriseOverYourHead opened the pull request:
# About the pull request

HackMD Link: [Here!](https://hackmd.io/@Pgile3jBTHSuq5x_dj0dnQ/Syoyr1YFxx)
Discord contributor-projects thread link: [Here!](https://discord.com/channels/150315577943130112/1409264665506353213/1409264665506353213)


This PR adds [tank desant](https://en.wikipedia.org/wiki/Tank_desant) mechanics to the M34A2 Longstreet Light Tank. Marines and xenos are now able to climb up and down atop the tank, as well as 'ride' with it as it moves through the map.

Almost all xeno ranged abilities still work. You can also shoot mobs atop the tank. Projectile behavior has been changed as of commit c790fa9655785abc65021a1bb3815fcc92a7807a to prioritize mobs riding ontop, and then hit the tank only if a target isn't found. Meat-shield infantry meta ✅. 

Tank has also been refactored to be multi-z friendly: The tank gunner and driver can both look one Z-level above. Edge cases where mobs fall down/climb down onto the tank from a higher Z are also taken care of.

Purchasing the ARC still disables the tank_crewman slot and overrides the vehicle ASRS terminals, but it is now possible to buy it even after the tank has been selected. This shouldn't be a problem, as it is expected that tankers will take all of their gear out at round start.

Dropped items such as weapons, grenades, ammo, et-cetera also inherit the 'Moving with the tank' mechanic from mobs. Your gun won't be left behind on the floor if you get stunned. This also works for deployed rollerbeds and stasisbags, for corpsman quality-of-life.

Lastly, marines and xenos alike atop the tank enjoy certain benefits and certain downsides. For example, the tank can rush into an incendiary OB (After it lands, obviously!) and have marines ontop pull wounded/dead marines to the top of their vehicle by securing a neck grab or a carry grab. However, dodging boiler shots, vanguard hooks or even pounces becomes a lot harder. Falling off of the tank is also really unpleasant!

Every new line of code has been tested and all new procs have been documented in accordance with the guidelines. Let me know if you have any doubts given the extensive nature of this PR.

<details> 
 <summary>Full list of changes</summary>
 
- Items can be thrown across the tank.
- Items thrown atop the tank will move and turn alongside with it.
- Items dropped atop the tank will also move and turn alongside with it.
- Roller beds, stasis bags, body bags, exceptionally, as structures, can be brought up and down the tank.
- Mobs firing wielded guns atop the tank get an accuracy and scatter debuff while it is moving. (('While it is moving' here means the debuff remains for around 3-4~ seconds after it has last moved.)) Pistols are unaffected.
- Mobs riding atop the tank get their sprites' pixel_y shifted up by 12 units.
- Tank now collides with but does not crash into open_space tiles, to prevent edge cases related to the tank falling down and being unrecoverable.
- New layer definition added: TANK_RIDER_LAYER at 4.51. (Tank chassis is at 4.1 and tank turret is at 4.5)
- Multitile/tank now allows colliding mobs to climb atop with climb_onto(), and climbing down with climb_down()
- The tank does not act as a dense object for mobs marked atop it.
- Crashing into a dense object at sufficient speed will jostle all mobs ontop and apply a 1.5 second WEAKEN.
- **Being forcefully dismounted from atop the tank by being grabbed, crashing, or thrown off applies a 1.5 second WEAKEN to humans, along with a 3 second superslow and a 5 second slow. (This is inconsistent with the video, which was made on an older version, but has been fixed.)**
- Whenever the tank invokes relaymove(), mobs ontop get their positions updated accordingly.
- Mobs riding atop the tank do not trigger facehugger traps.
- **Mobs atop the tank STILL TRIGGER normal acid traps. (This is inconsistent with the video, which was made on an older version, but has been fixed.)**
- Adds a myriad of documented helper functions to assist with this new kind of movement.
- Living.dm code has been updated to handle these new types of interactions with the tank.
- Mobs atop the tank do not get burnt by flames below.
- Mobs atop the tank do not trigger boiler's foul resin trap.
- Mobs atop the tank do not get damaged by stepping into resin spikes.
- Mobs can be dragged atop the tank by securing a grab and calling the climb_onto function (Like how you can help others up window frames)
- Mobs can be dragged atop the tank by securing a GRAB_CARRY or a GRAB_CHOKE, if the mob pulling them is already ontop the tank. 
- Mobs who die atop the tank get layered atop it until they are not atop the tank anymore.
- Xeno acid spray is now layered at TANK_RIDER_LAYER
- Xeno acid spray is not stopped by multitile/tank turfs anymore.
- Oppressor Prae's ambush, xenomorph pounces, warrior's lunge, boiler gas and trapper mines **all still work on mobs atop the tank**. 
- Xenos with movement abilities such as the runner's pounce can use those abilities to get up or down the tank with no delay.
- Mobs atop the tank do not trigger claymores.
- Mortar, M2C and deployable cades cannot be deployed on a tank tile.
- Xenos attacking the tank from ontop with SLASH attacks get a 30% damage penalty and cannot hit the treads module.
- The ARC can now be bought even after the tank has been brought up. Doing this will override the tank's modules and the tank_crew access on the ASRS console, but this shouldn't be a problem given the tank will most likely have already been brought up and outfitted.
- **ARC still disables the tank and tank_crew slots if it is purchased before the tank is brought up and outfitted.**
- Hardpoint weapons slightly tweaked to allow and include scatter when firing on a different Z
- The penalty for being run-over by the tank has been changed: You now receive a SLOW effect for the same duration of the old knockdown, but the knockdown has been halved.
- Marines can crawl underneath the tank by using the 'rest' verb and moving.

Please point it out if I've forgotten anything!
</details>


# Explain why it's good for the game

Combined arms warfare in CM is a really interesting concept with a poor implementation. It's my understanding that this comes from the fact that the tank behaves as a 3x3 impenetrable wall. It blocks pushes. It blocks retreats. It blocks firing lanes. It blocks body recovery. It's a SLOG to play alongside with. This PR aims to solve that.

By fixing the number 1 issue that led to the tank's removal as a mainstay of CM, this PR aims to be the first step into re-adding the tank into normal matches, paving the way towards a state where it can be more properly balanced and played as, with, and against.

As the tank is still primarily in a fire support role (No LTB or armor), I hope that the impact on balance will be somewhat minimal. And while riding atop the tank can be (and should be!) used to coordinate a push, I think xeno players will also enjoy how much more vulnerable being on top of it makes you. Boilers and Oppressors, especially, will be that much more relevant.

I will fix bugs and apply changes as they are requested.

# Testing Photographs and Procedure
<details>
<summary>Testing Photographs and Procedure</summary>

Every modified line of code was filled with printf()-esque statements to ensure they are being ran correctly.

A comprehensive amount of edge cases have been tested, such as but not limited to: Climbing on the tank while it is moving, falling onto the tank from a higher Z-level, shooting up at targets above Z-levels, almost ALL xeno abilities atop tank riders, et-cetera...

Here's a small snippet of those such tests:

<img width="1363" height="668" alt="Capture2" src="https://github.com/user-attachments/assets/296ee2fc-29a7-42ba-ab79-84764a2013c2" />
</details>

# Screenshot & Videos
<details>
<summary>Screenshots & Videos</summary>

<img width="377" height="374" alt="Capture" src="https://github.com/user-attachments/assets/f9b95c35-5289-467d-817e-9422ca9a1696" />

There's also an older video in the discord thread [here!](https://discord.com/channels/150315577943130112/1409264665506353213/1409264724956151868)

</details>


# Next Steps

- Resin constructions interactions: Sticky resin could slow down the tank, resin spikes could apply damage to the treads, et-cetera. In general, more anti-vehicle tools for xenos are something I am looking into implementing if this PR is well received.

- Module Repairs: Hardpoints could be changed to be repairable instead of being Q-DEL'd once their health reaches zero. This is also something I am looking into implementing on another PR.

- Pop cap: These changes address the main issue that led to the tank being removed from normal gameplay rotation. Hopefully the pop cap can be reduced in the future (120-150 territory?) to allow for more tank rounds.

- Thwomper has expressed interest in seeing those mechanics ported over to V-TOL. Those functions can be made part of the vehicle/multitile class without much effort, but a better idea might be making a vehicle/multitile/rideable class.

# Special Thanks:

- **4khan** for giving me some encouragement and steering me into the right direction as to who to approach with this.
- **MrDoritosMan** for some invaluable help testing edge-cases with me in a private server
- **Everyone in the contributor-projects thread** for expressing their ideas and opinions in a mature, constructive manner.
- **YOU**, for taking the time to look at this, and for adding **technical** comments to this PR, and **non-technical** comments to the [contributor-projects thread](https://discord.com/channels/150315577943130112/1409264665506353213/1409264665506353213).

# Changelog
:cl: BugWasabi
add: Marines are now able to ride atop the M34A2 Light Tank, tank-desant style.
add: Items now inherit the same desant mechanic from mobs.
balance: The ARC can now be purchased even after the TANK has been sent up.
code: Xeno abilities and projectile code tweaked to allow targeting of mobs atop the tank.
code: Xenomorph castes with pounce-like abilities can use them to instantly climb up or down the tank.
/:cl:

SunriseOverYourHead committed:
Tank Desant: Rideable Tank And Overhauled Interactions

cuberound commented:
not sure if not done already but perhaps look into preventing tank from going to lower z/ make it able to travarse back up, as there were issues with tank falling to lower level on tyrargo and being unable to get out, I think

SunriseOverYourHead committed:
Tank Desant: Collide with open_space turfs

cuberound commented:
> > not sure if not done already but perhaps look into preventing tank from going to lower z/ make it able to travarse back up, as there were issues with tank falling to lower level on tyrargo and being unable to get out, I think
> 
> A very valid edge case.
> 
> I'm going to make the tank collide with (but not crash) into open_space turfs pending a more refined solution...
> 
> ...Ideally a stairs subtype (A ramp, if you will) that allows multitile movement up and down Zs would solve this issue, but I'd also need to revisit transform maths to include mobs changing Z-levels, and edge cases where the tank is partially on a Z, and partially on another. This would probably be a different PR on its own.
> 
> // Done on commit [370d755](https://github.com/cmss13-devs/cmss13/commit/370d7551d640fce129a30e74318981d1d8f0bc3a)

when I think of it something of that sort will prop come in the multiz ds rework, so no sweat

cm13-github commented:
<!-- tgs_test_merge_comment -->
## Test merge deployment history:

<details><summary>Test Merge Deployed @ 09/21/2025 17:34:56 +00:00:</summary>

##### Server Instance

[CM13 Live](https://github.com/cmss13-devs/cmss13/deployments/activity_log?environment=TGS%3A+CM13+Live)

##### Revision
Origin: fc382810ca0a66cb5e04c7bfcde3747836fb1277
Pull Request: 370d7551d640fce129a30e74318981d1d8f0bc3a
Server: 540e85df744a0d74b7a16ea9f7924e511273c55d
</details>
<details><summary>Test Merge Removed @ 09/21/2025 18:37:01 +00:00:</summary>

##### Server Instance
CM13 Live
</details>
<details><summary>Test Merge Deployed @ 09/23/2025 18:26:42 +00:00:</summary>

##### Server Instance

[CM13 Live](https://github.com/cmss13-devs/cmss13/deployments/activity_log?environment=TGS%3A+CM13+Live)

##### Revision
Origin: 4b0dbdf5d66f3cc04988b17836605d4e328690b7
Pull Request: 09580281dae59dd6834126304f3cd522da265b7a
Server: 3bcb72b376195e03ab1c4e335acdfac408ef983b
</details>
<details><summary>Test Merge Removed @ 09/23/2025 20:56:56 +00:00:</summary>

##### Server Instance
CM13 Live
</details>
<details><summary>Test Merge Deployed @ 11/01/2025 23:40:13 +00:00:</summary>

##### Server Instance

[CM13 Live](https://github.com/cmss13-devs/cmss13/deployments/activity_log?environment=TGS%3A+CM13+Live)

##### Revision
Origin: ce9c5e90af881778d96fcac81baff6b3112925dd
Pull Request: 3575bc9928e64294978520f02010986726683593
Server: 8fb757d61971e3565181bf32c6f8987d43250891
</details>
<details><summary>Test Merge Removed @ 11/02/2025 00:59:25 +00:00:</summary>

##### Server Instance
CM13 Live
</details>
<details><summary>Test Merge Deployed @ 12/22/2025 01:52:37 +00:00:</summary>

##### Server Instance

[CM13 Live](https://github.com/cmss13-devs/cmss13/deployments/activity_log?environment=TGS%3A+CM13+Live)

##### Revision
Origin: 35c0ab854d9e1f13e9a0d8f63d78ecc8f8c1420e
Pull Request: ba4b049bc7511bd8b612f46f7a4510bf0a4bbcff
Server: 2424d2c0f7646549a66fd999a0929de8940b4a92
</details>
<details><summary>Test Merge Removed @ 12/22/2025 05:12:34 +00:00:</summary>

##### Server Instance
CM13 Live
</details>
<details><summary>Test Merge Deployed @ 05/17/2026 17:25:58 +00:00:</summary>

##### Server Instance

[CM13 Live](https://github.com/cmss13-devs/cmss13/deployments/activity_log?environment=TGS%3A+CM13+Live)

##### Revision
Origin: 97dc050ecd140e7e6a731400954736e58c36ca27
Pull Request: af663eea2a9df9f38bc3cd4ba934df27654f00d5
Server: a6e7bebc042f540bc8648623c6a551fa978b4658
</details>
<details><summary>Test Merge Updated @ 05/17/2026 17:57:55 +00:00:</summary>

##### Server Instance
CM13 Live

##### Revision
Origin: ec252dc031e34a7b10c66d98247254b70de2f70a
Pull Request: af663eea2a9df9f38bc3cd4ba934df27654f00d5
Server: bb5f2377ff79f91ec8286acdde578599ff61a753
</details>
<details><summary>Test Merge Removed @ 05/17/2026 18:23:08 +00:00:</summary>

##### Server Instance
CM13 Live
</details>

SunriseOverYourHead committed:
Reduces mob stun time from medium vehicle collision

SunriseOverYourHead committed:
Marines can crawl away from underneath the tank

This is an important fix to an edge case where a marine might get stuck in the middle tile of the 3x3 Tank and be unable to get out. Now they can simply crawl away from underneath the Longstreet tank. Tested thoroughly with other grab interactions.

SunriseOverYourHead committed:
Entities atop the tank now get a pixel_y offset

SunriseOverYourHead committed:
Being pulled off the tank is less punishing

The whole idea of receiving a debuff when you're pulled off of the tank is to penalize drivers who use the tank to spearhead a push and allow xenos to take advantage of poorly positioned marines atop the tank. A 3 second weaken is a bit too punishing and basically guarantees a cap. Now, the marine will get a shorter stun and a hefty slowdown, but they'll be able to fight back and can be saved if other marines pay attention.

SunriseOverYourHead committed:
WIELDED weapons accuracy debuff when fired atop moving tank

SunriseOverYourHead committed:
TANK DESANT now wont block VTOL open_space movement

SunriseOverYourHead committed:
removes debug statements from gun.dm

Oopsie daisy, almost forgot.

SunriseOverYourHead committed:
improves climb_down logic

Checks if we have a valid turf to climb down before climbing down, otherwise, we'd 'begin climbing down' even if we tried to disembark onto a wall. (Which would obviously fail, but, why allow it to start in the first place?)

SunriseOverYourHead committed:
Projectile interaction changes upon colliding with TANK

Projectiles now by default go 'over' the tank when crossing tank tiles, automatically hitting mobs on top. Sprite clicking is not necessary anymore. If the projectile does not find a valid target atop the tank, it will automatically hit the tank upon attempting to leave turfs with a tank tile.

You still can not shoot entities opposite from you on the tank.

This will make it easier for xenos to land their ranged abilities, and the hedgehog Rav's shield, bone spurs, and other such abilities have been tested to be working correctly.

As projectiles now try to hit mobs ontop first... Meatshield infantry meta?

SunriseOverYourHead committed:
Edge case fix: Dropping hauled mob atop tank

If a xenomorph walks atop the tank with a hauled mob, its sprite will now be layered properly atop the tank.

Moreover, releasing a hauled mob atop the tank now correctly marks such a mob as being on top of said tank.

SunriseOverYourHead committed:
Item Desant: Part 1

Items can be thrown across the tank, even if you're not atop it. Items can now also move alongside the tank when dropped or thrown atop. Exceptionally, roller beds, cryo bags and bogybags (deployed versions) can also be brought up and down the tank despite being obj/structures. Further refinement of the procs that handle transitioning between the tank top and down with a dragged mob.

TODO: Handle edge case for roller beds with stasis or bodybags ontop.

SunriseOverYourHead committed:
Improves roller beds and stasis bags on tank

SunriseOverYourHead committed:
Optimizes rider jousting and rider motion

SunriseOverYourHead committed:
Fixes item layering bug when dropped atop tank

SunriseOverYourHead commented:
Proof of test of commit fad6163e631a581975031a482817e3e497f86364:
https://discord.com/channels/150315577943130112/1409264665506353213/1436551142963413032

Proof of test of commit cf6367893b1be013337651766676a6a0409e1546:
https://discord.com/channels/150315577943130112/1409264665506353213/1436551473344544799

Proof of test of commit 9097142fde7b75764690de511ee5df3f93e89791:
https://discord.com/channels/150315577943130112/1409264665506353213/1436551653942759454

SunriseOverYourHead committed:
Fixes mobs atop not falling down

Fixes an oversight caused by cf6367893b1be013337651766676a6a0409e1546

SunriseOverYourHead committed:
medevac rollers can't deploy atop tank

SunriseOverYourHead committed:
resin doors now properly block tank movement

SunriseOverYourHead committed:
climb up and down time tweaks.

Xenos now climb down faster and take longer to climb up. Marines now climb up faster but still take a while to climb down. The speed of climbing up and down now depends on vehicle momentum.

SunriseOverYourHead committed:
prevents xenos from hiding atop the tank.

SunriseOverYourHead committed:
Adds more tank riding layers

SunriseOverYourHead committed:
runner pounce fix

fixes runners being dismounted improperly (or not at all) from pounces from atop the tank

SunriseOverYourHead committed:
tank now goes over cargo trolleys

SunriseOverYourHead committed:
penalty from being run over now scales with momentum

SunriseOverYourHead committed:
Tank momentum and jousting refactor

SunriseOverYourHead committed:
Linters: space indentation fix

SunriseOverYourHead committed:
Removes leftover debug statements

woops!

SunriseOverYourHead committed:
fixes movement collision oversight

SunriseOverYourHead committed:
fixes movement collision oversight

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

SunriseOverYourHead committed:
Merge branch 'tankFixes' of https://github.com/SunriseOverYourHead/cmss13 into tankFixes

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

SunriseOverYourHead committed:
Linters Fix

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

Never fixing merge conflicts through github again.

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

SunriseOverYourHead committed:
First round of linter fixes

Red-byte3D commented:
code: The ARC can now be purchased even after the TANK has been sent up.
 balance change, please properly add tags

SunriseOverYourHead commented:
> code: The ARC can now be purchased even after the TANK has been sent up. balance change, please properly add tags

Wasn't sure if it was CODE or BALANCE. It's been changed and applied.

SunriseOverYourHead committed:
Prevents tank riders from being FF'd by non IFF projectiles.

SunriseOverYourHead committed:
Barricades no longer 'HARD CRASH' the tank.

Nor do they get destroyed when running over them IF you don't attempt to cross over them. (Aka, you can drive the tank flush to cades from the inside side without destroying them.)

SunriseOverYourHead commented:
Proof of testing of commits 5e4690e8b5fedeb5c8763861fc79e6a9d8e9f589 and 532aec69d4318c9aef80c7056f4b8dbf6a040145:

https://discord.com/channels/150315577943130112/1409264665506353213/1505984176078327819



> A few notes (some of these might be intended, unsure) from TM round 31515 as driver:
> 
>     * If you drive over a cade facing the same direction as the tank, it will push back marines on top of the tank without destroying the barricade if it's not at the tank's center.
> 
>     * There appeared to be a few ledge oddities, but I wasn't able to reliably replicate them.
> 
>     * Moving or rotating the tank while actions are taken on top of it interrupts those actions, including trying to dismount the tank itself. This was somewhat inconsistent, with players sometimes being able to get on/off while moving and sometimes not.
> 
>     * Opening a barricade while driving into it triggers a crash and launches marines/items, even if nothing else was hit or otherwise damaged
> 
>     * Trying to enter areas the tank can't enter (e.g. staircases) triggers a crash.
> 
>     * There was one instance of a human on top of the tank blocking tank weapon fire (A GL which was blocked by a VAISO ERT member standing in the center of the tank's roof).

This fixes the barricade issues and the last issue regarding the GL.

Git-Nivrak commented on code/__DEFINES/layers.dm:
```diff
@@ -138,6 +138,18 @@
 #define WEATHER_LAYER 4.14
 
+/// For objs atop a tank
+#define TANK_RIDER_OBJ_LAYER 4.505
```
Move weather up to be above all this

Git-Nivrak commented on code/game/objects/effects/aliens.dm:
```diff
@@ -43,7 +43,7 @@
 	opacity = FALSE
 	anchored = TRUE
-	layer = ABOVE_OBJ_LAYER
+	layer = TANK_RIDER_LAYER
```
Why?

SunriseOverYourHead replied:
This is required to stop the acid spray attack (Not the projectile, the multi-tile spray) from boilers/praes/spitters from layering underneath the tank.

After I modified the acid spray to go 'through' the tank and hit multiple tiles (With the intent of affecting multiple marines atop a tank), I noticed that the spray decal was being layered underneath its sprite. This layer change prevent that.

SunriseOverYourHead replied:
* It also affects oppressor prae hooks, more critically, which can also target marines atop the tank. The sprites marking its area of effect get obscured by the tank sprite otherwise.

Git-Nivrak reviewed (changes requested):
Before I go through this fully this you need to refactor `tank_on_top_of` and related procs into a component which will hold all the relevant vars
Also you should probably turn the tank reference into a weakref

There are also a lot of one letter vars and src usage, please make sure to go through the contributing guidelines so it is easier for the PR to be reviewed and approved

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

SunriseOverYourHead committed:
one letter var removals, changes for clarity, tank_on_top weakref

SunriseOverYourHead committed:
weather layer increased

SunriseOverYourHead committed:
forgot the .dme

SunriseOverYourHead commented on code/__DEFINES/layers.dm:
```diff
@@ -132,8 +132,21 @@
+/// For objs atop a tank
+#define TANK_RIDER_OBJ_LAYER 4.505
+/// Mostly for bodybags atop a tank
+/// Not used yet. This is a reminder for me to add a specific tank_layer var to OBJs to keep track of
```
It's been a while since I wrote this comment and I remember having in mind a different idea to sort through layer orderings between objs. Something about adding in a new var 'tank_layer_add' to each object, and then adding it to an OBJ's layer.

At the moment, however, I don't feel like this would be a better idea than just going with the current setup.

SunriseOverYourHead commented on code/_onclick/human.dm:
```diff
@@ -156,6 +156,23 @@
+				tank.mark_on_top(target)
+				target.update_transform(TRUE)
+				return
+
```
The idea behind this whole block is allowing marines to aggro-grab and pull others atop the tank. Kind of how you can grab people to pull them down stairs, or into a window frame.

SunriseOverYourHead commented on code/game/objects/effects/aliens.dm:
```diff
@@ -43,7 +43,7 @@
 	opacity = FALSE
 	anchored = TRUE
-	layer = ABOVE_OBJ_LAYER
+	layer = TANK_RIDER_LAYER
```
This needs to exist, because otherwise, skills like prae hooks end up layered UNDER the tank sprite, making it very hard to dodge.

SunriseOverYourHead commented on code/game/objects/items/bodybag.dm:
```diff
@@ -83,6 +91,7 @@
 	/// How many extra pixels to offset the bag by when buckled, since rollerbeds are set up to offset a centered horizontal human sprite.
 	var/buckle_offset = 5
 	store_items = FALSE
+	is_allowed_atop_vehicle = TRUE // Allows bodybags, exceptionally, as structures, to be used atop vehicles such as the tank.
```
Same thought process for stasis bags and roller beds, but I will only make this comment once:

The intent isn't to allow surgery atop the tank, as some players commented. Rather, this is just corspman QoL so they can just cross over the tank while dragging a body/stasisbag or rollerbed instead of:

Unbuckle a dragged mob
Fold stasis bag and bed.
Drag mob atop the tank
Cross to the otherside
Dismount
Deploy bed/stasis bag
Buckle mob back onto it
Carry on towards FOB/ETC.


...Besides, surgery steps get interrupted if the tank moves, anyhow.

SunriseOverYourHead commented on code/game/objects/structures/barricade/deployable.dm:
```diff
@@ -147,6 +147,12 @@
+		if(L.is_on_tank_hull())
+			to_chat(user, SPAN_WARNING("[singular_name] cannot be built here!"))
+			return
+
```
There is nothing that stops cades from actually 'working' atop the tank. They just arbitrarily can't be dragged or built atop.

 And while it could pay homage to how soldiers used to stack sandbacks, treads, etc... atop their tanks (Like sherman crews in WW2!), it also felt like it'd be a tad too OP, or just silly.

SunriseOverYourHead commented on code/game/objects/structures/stool_bed_chair_nest/bed.dm:
```diff
@@ -94,7 +104,10 @@
 	if(buckled_bodybag)
-		buckled_bodybag.pixel_y = initial(buckled_bodybag.pixel_y)
+		if(is_atop_vehicle())
+			buckled_bodybag.pixel_y = initial(buckled_bodybag.pixel_y) + 12 // Magic number bad. TODO: Make tank pixel offset its own variable.
```
Yeah, it's bad, but, in hindsight, it's only being used on three places.

SunriseOverYourHead commented on code/game/turfs/open_space.dm:
```diff
@@ -105,7 +105,18 @@ GLOBAL_DATUM_INIT(openspace_backdrop_one_for_all, /atom/movable/openspace_backdr
+	if(tank_at_destination && isliving(user))
+		var/mob/living/L = user
+		tank_at_destination.mark_on_top(L)
+
```
These just basically make it so that if you fall atop a tank from a higher Z-level, you'll be placed ontop the tank by default, rather than 'clipping' through it.

SunriseOverYourHead commented on code/modules/cm_aliens/structures/trap.dm:
```diff
@@ -94,6 +94,10 @@
+				// tank treads block the hugger's hole. This behavior can be also added to normal acid traps.
+				var/obj/vehicle/multitile/tank/victim_tank = victim_mob.get_tank_on_top_of()
+				if(victim_tank && (get_turf(victim_mob) in victim_tank.locs))
+					return
```
The idea being that it isn't particularly fun for marines if facehuggers clip through the tank to hug them. This behavior was also applied to acid traps before, in a previous version of Desant, but I feel like a heavy vehicle walking atop a resin hole would definitely cause it to 'splash'.


Granted, the tank is also not in a state where it can spearhead a push and run through traps without dooming itself.

SunriseOverYourHead commented on code/modules/cm_aliens/structures/trap.dm:
```diff
@@ -109,6 +113,10 @@
+					// just in case a Greeno decides to ride on the tank.
+					var/obj/vehicle/multitile/tank/victim_tank = victim_xeno.get_tank_on_top_of()
+					if(victim_tank && (get_turf(victim_xeno) in victim_tank.locs))
+						return
```
This is a leftover from when acid traps were still not triggered for mobs atop the tank.

The idea was handling an edge case where a greeno is atop the tank but triggers an acid trap below it.

I feel like this is mostly dead code since they can't get facehugged by another hive.

SunriseOverYourHead commented on code/modules/cm_aliens/XenoStructures.dm:
```diff
@@ -408,7 +412,7 @@
 
 /obj/structure/mineral_door/resin/proc/take_damage(dam, mob/mob)
-	health -= dam
+	update_health(dam)
```
I genuinely forgot why I did this. I think -=dam was having some kind of undesirable behavior, which update_health didn't have.

SunriseOverYourHead commented on code/modules/cm_marines/m2c.dm:
```diff
@@ -111,6 +111,11 @@
+		var/mob/living/L = user
+		if(L.is_on_tank_hull())
+			to_chat(user, SPAN_WARNING("You can't set up \the [src] here."))
+			return FALSE
```
Very good idea as a vehicle module. Not so good as something that can be freely deployed atop the Longstreet.

SunriseOverYourHead commented on code/modules/cm_tech/techs/marine/tier1/arc.dm:
```diff
@@ -16,11 +16,14 @@
+	//	var/obj/structure/machinery/cm_vending/gear/vehicle_crew/gearcomp = GLOB.VehicleGearConsole
+	//	if(gearcomp.selected_vehicle == "TANK")
+	//		to_chat(unlocking_mob, SPAN_WARNING ("A vehicle has already been selected for this operation."))
+	//		return FALSE
```
The idea being that the ARC and the TANK have entirely different niches, and shouldn't be mutually exclusive.

This could be more refined, though: Buying the ARC shouldn't make you unable to select other tank modules if you still have unselected hardpoints. But this would require either remapping the Almayer to have two distinct vehicle consoles, or doing a fairly major refactor of how the vending machine overlay functions. 

(Probably adding a header component that allows you to cycle between 'bought' vehicles and their hardpoints.)

SunriseOverYourHead commented on code/modules/mob/living/carbon/xenomorph/abilities/ability_helper_procs.dm:
```diff
@@ -301,7 +301,8 @@
 		qdel(temp)
-		if(blocker)
+		// despite the tank being a blocker, acid sprays should still go through it to hit marines ontop.
+		if(blocker && !locate(/obj/vehicle/multitile/tank) in turf.contents)
```
Without this, acid sprays stop at the first tank tile they hit.

With this, it goes on to hit marines ontop and even behind the tank.

SunriseOverYourHead commented on code/modules/mob/living/carbon/xenomorph/abilities/general_abilities.dm:
```diff
@@ -408,20 +408,25 @@
+	remove_hide_status()
+	apply_cooldown(4) //2 second cooldown after attacking
+
+/datum/action/xeno_action/onclick/xenohide/proc/remove_hide_status()
```
This entire add diff blocks huggers and runners from hiding under the tank WHILE atop it and getting free hugs/slashes against marines.

SunriseOverYourHead commented on code/modules/mob/living/carbon/xenomorph/XenoProcs.dm:
```diff
@@ -233,20 +233,35 @@
+		victim.layer   = initial(victim.layer)
+		victim.plane   = initial(victim.plane)
+		victim.pixel_y = initial(victim.pixel_y)
+
```
When I added this, I considered the fact that some players might try to plug holes with the tank.

It'd get destroyed. And then runners/xenos would have a hard time dragging caps over it and back into the hive.

So, this ensures that, despite being a rare edge case, mobs being dragged by other xenos are still handled correctly even if they cross tank tiles. I think that undevouring/unhauling marines atop the tank also correctly flags the marine as being atop the tank.

SunriseOverYourHead commented on code/modules/mob/living/living.dm:
```diff
@@ -156,6 +156,76 @@
+ *   - They use some kind of skill to dismount (e.g pouncing outside)
+ *   - They wait 1 second until climb_down finishes.
+ *
+ * OOPS! As a fix for LINTERS, climb_down() is called on the proc in carbon.dm
```
I need to remove this, as the fix for this specific linters issue (Which is still fresh on my mind today because it was such a goose-egg hunt to fix) was to actually add a separate proc for carbon simplemobs.

SunriseOverYourHead commented on code/modules/mob/mob_grab.dm:
```diff
@@ -116,6 +116,25 @@
+				tank.mark_on_top(victim)
+				victim.update_transform(TRUE)
+				return
+
```
Part of the many shotgun-style changes to all grabbing systems I had to do to ensure marines, xenos, items can all be dragged atop.

SunriseOverYourHead commented on code/modules/mob/mob_verbs.dm:
```diff
@@ -280,6 +295,10 @@
+	// this is probably a bit hacky.
+	if(!(vehicle_turf == null))
+		above = locate(vehicle_turf.x, vehicle_turf.y, vehicle_turf.z+1)
+
```
The idea behind the 3 add diffs in this file is allowing the tank gunner and driver to look up at the multi-Z level above the tank.

SunriseOverYourHead commented on code/modules/vehicles/multitile/multitile.dm:
```diff
@@ -53,7 +53,13 @@
 	// Determines how much slower the vehicle is when it lacks its full momentum
 	// When the vehicle has 0 momentum, it's movement delay will be move_delay * momentum_build_factor
 	// The movement delay gradually reduces up to move_delay when momentum increases
+	//// This is mostly vestigial and it looks like it cancels itself out when calculating movement delay??? -bwsb
```
This entire file; along with multitile_bump and ESPECIALLY multitile_movement are largely **hacky and bad** and are deserving of a full re-work.

Trying to make sense of how movement is handled by vehicles while working on Desant, especially the nonsensical formulas with momentum, and how abstract figuring out a vehicle's top speed and acceleration is, was one of the major hurdles with this PR.

Knowing this, everything here was tested extensively with multiple vehicle types.

SunriseOverYourHead commented on code/modules/vehicles/multitile/multitile_movement.dm:
```diff
@@ -140,12 +141,12 @@
 /obj/vehicle/multitile/proc/update_next_move()
 	// 1/((m/M)*b) where m is momentum, M is max momentum and b is the build factor
+	//// move_momentum_build_factor seems to cancel itself out here. It's worth to revisit this section and maybe refactor it.
 	var/anti_build_factor = 1/((max(abs(move_momentum), 1)/move_max_momentum) * move_momentum_build_factor)
```
Nonsensical formula.

On line 147, movement_build_factor cancels itself from line 145.

As per my last comment on multitile.dm being hacky and bad.

SunriseOverYourHead commented on code/modules/vehicles/multitile/multitile_movement.dm:
```diff
@@ -307,3 +333,59 @@
+ * That way, all of the effects that happen when crashing are in the same spot.
+ */
+/obj/vehicle/multitile/proc/on_crash()
+	return
```
I wrote this and seldom used it. Still, instead of separating crash behavior by light, medium or heavy vehicles, it'd be better to have a specific on_crash proc with them.

The idea was that... Since we HAD V-TOL being worked on, I figured that vehicle 'weight' classes were obsolete and it'd be better to define on-crash behavior for each specific vehicle.

SunriseOverYourHead commented on code/modules/vehicles/multitile/multitile_movement.dm:
```diff
@@ -307,3 +333,59 @@
+	if(abs(move_momentum) > 0 && !momentum_decay_active)
+		momentum_decay_active = TRUE
+		spawn(0)
+			momentum_decay_loop()
```
Momentum in Tank Desant is slightly changed and mostly used to calculate firing delay penalties for marines atop.

Two important scenarios:
- Tank is standing flush behind a barricade, holding FOB:
Marines should  be able to seamlessly climb it and fire from it with no penalties.
- Tank is darting around the frontline and spearheading a push:
Unstable surface. Lots of momentum. Marines should have acc penalties.

SunriseOverYourHead commented on code/modules/vehicles/tank/tank.dm:
This file along with tank_helpers has the majority of the 'hard to review' code.

Still, I hope the new procs here are documented well enough to facilitate reviewing.

SunriseOverYourHead commented on code/modules/vehicles/tank/tank_helpers.dm:
Ditto, from the comment on tank.dm.

The helper.dm file was made halfway through when I considered the possibility of adding desant behavior to other vehicles, instead of keeping it locked to the tank.

Doing this now would be a bit of a headache, mostly because I'd have to port the changes in tank.dm to vehicles.dm, but it's definitely not all that bad thanks to tank_helpers being its own separate file.

SunriseOverYourHead commented on code/modules/vehicles/tank/tank.dm:
```diff
@@ -165,42 +195,439 @@
-/obj/vehicle/multitile/tank/MouseDrop_T(mob/dropped, mob/user)
-	. = ..()
-	if((dropped != user) || !isxeno(user))
+// !!!! No point in keeping this now that you can freely climb onto the tank. !!!!
```
Evaluating whether or not a xeno might end up in the following edge case:

1) Gets stunned over  by a HEDP or another stun source
2) Gets ran over by Tank while stunned
3) Tank breaks while the stunned xeno is at the MIDDLE tile of the 3x3 vehicle
4) Can't click to climb out of. Can't reach an edge to 'climb atop'. Can't 'lie down and crawl away' like humans can.

SunriseOverYourHead commented on code/modules/vehicles/tank/tank.dm:
```diff
@@ -165,42 +195,439 @@
+			return
+
+	// Humans and preds climb up faster. Humans can't climb down instantly.
+	// takes between 0.25 seconds when still and 1.20 second when at full speed (You'd have to climb through one of the sides)
```
For xenos and humans;
Both values empirically tested with the following things in mind:

1) Tank is a marine asset. Marines shouldn't get punished for playing around the tank.
Thus - Faster climbing up. Non-instant climbing down.

2) 'Trapping' xenos atop the tank if they happen to climb up shouldn't lead to a cheesy kill.
Thus - Xenos can climb down almost instantly. Xenos take longer to climb up to avoid accidentally going ontop and becoming an easy target, but not so long that it is viable to block chokes with the tank.

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.

SunriseOverYourHead committed:
fix linter

SunriseOverYourHead commented:
The second PR steps on the toes of this one, and the third PR uses the second PR as a rug with how often it re-touches the same files multiple times.

I am most likely going to close this and re-open everything in a single big, atomized PR.
Let me know if someone is currently reviewing this. In that case, I'll keep it open.

cm13-github commented:
This pull request has conflicts, please resolve those before we can evaluate the pull request.

SunriseOverYourHead committed:
Merge branch 'master' into tankFixes

cm13-github commented:
Conflicts have been resolved. A maintainer will review the pull request shortly.
